graphql_compiler/query_formatting/common.py

```diff
@@ -90,27 +94,36 @@ def _deserialize_anonymous_json_argument(expected_type: GraphQLScalarType, value
     if not isinstance(value, expected_python_types):
         correct_type = False
     if not correct_type:
-        raise ValueError(
-            "Unexpected type {}. Expected one of {}.".format(type(value), expected_python_types)
-        )
+        raise _raise_invalid_type_error(name, expected_python_types, value)
+
+    custom_deserialization_functions = merge_non_overlapping_dicts(
+        {
+            # Bypass the GraphQLFloat parser and allow strings as input. The JSON spec allows only
+            # for 64-bit floating point numbers, so large floats might have to be represented as
+            # strings.
+            GraphQLFloat.name: float,
+            # Bypass the GraphQLInt parser and allow long ints and strings as input. The JSON spec
+            # allows only for 64-bit floating point numbers, so large ints might have to be
+            # represented as strings.
+            GraphQLInt.name: int,
+        },
+        {
+            # Since we cannot serialize the parse_value function of custom scalar types when
+            # serializing a schema, it is possible that the parse_value is incorrectly set. We must,
+            # therefore, use the parse_value function of the original scalar type definition.
+            graphql_type.name: graphql_type.parse_value
+            for graphql_type in CUSTOM_SCALAR_TYPES
+        },
+    )
 
-    name_to_custom_type = {graphql_type.name: graphql_type for graphql_type in CUSTOM_SCALAR_TYPES}
-    # Bypass the GraphQLFloat parser and allow strings as input. The JSON spec allows only for
-    # 64-bit floating point numbers, so large floats might have to be represented as strings.
-    if is_same_type(expected_type, GraphQLFloat):
-        return float(value)
-    # Bypass the GraphQLInt parser and allow long ints and strings as input. The JSON spec allows
-    # only for 64-bit floating point numbers, so large ints might have to be represented as strings.
-    elif is_same_type(expected_type, GraphQLInt):
-        return int(value)
-    # Use the default GraphQL parser to parse the value
-    elif expected_type.name in name_to_custom_type:
-        # Since we cannot serialize the parse_value function of custom scalar types when
-        # serializing a schema, it is possible that the parse_value is incorrectly set. We must,
-        # therefore, use the parse_value function of the original scalar type definition.
-        return name_to_custom_type[expected_type.name].parse_value(value)
-    else:
-        return expected_type.parse_value(value)
+    deserialization_function = custom_deserialization_functions.get(
+        expected_type.name, expected_type.parse_value
```

Comment:
Under what circumstances is the `.get()` expected to go to the "backup" `expected_type.parse_value` option? Can we make those situations explicitly handled by the dict or otherwise, and raise `AssertionError` for everything else? I'm worried this kind of catch-all can cause difficult-to-track-down errors down the road.

Reply:
Addressed. Decided to go for a full refactor.